lib/elixir/lib/uri.ex

```diff
@@ -373,6 +373,77 @@ defmodule URI do
       "http://google.com"
   """
   defdelegate to_string(uri), to: String.Chars.URI
+
+  @doc ~S"""
+  Merges two URIs as per RFC 3986 §5.2
+
+  ## Examples
+
+      iex> URI.merge(URI.parse("http://google.com"), "/query") |> to_string
+      "http://google.com/query"
+
+      iex> URI.merge("http://example.com", "http://google.com")|> to_string
+      "http://google.com"
+  """
+  def merge(%URI{authority: nil}, %URI{}), do: raise ArgumentError, "you must merge onto an absolute URI"
```

Comment:
I think we should match on `_rel` instead of `%URI{}`, thus we don't even try to parse second argument when it's a binary.